In [4]:
import pandas as pd 
leads = pd.read_csv("Leads.csv")
leads.head()

,Prospect ID,Lead Number,Lead Origin,Lead Source,Do Not Email,Do Not Call,Converted,TotalVisits,Total Time Spent on Website,Page Views Per Visit,...,Get updates on DM Content,Lead Profile,City,Asymmetrique Activity Index,Asymmetrique Profile Index,Asymmetrique Activity Score,Asymmetrique Profile Score,I agree to pay the amount through cheque,A free copy of Mastering The Interview,Last Notable Activity
0,7927b2df-8bba-4d29-b9a2-b6e0beafe620,660737,API,Olark Chat,No,No,0,0.0,0,0.0,...,No,Select,Select,02.Medium,02.Medium,15.0,15.0,No,No,Modified
1,2a272436-5132-4136-86fa-dcc88c88f482,660728,API,Organic Search,No,No,0,5.0,674,2.5,...,No,Select,Select,02.Medium,02.Medium,15.0,15.0,No,No,Email Opened
2,8cc8c611-a219-4f35-ad23-fdfd2656bd8a,660727,Landing Page Submission,Direct Traffic,No,No,1,2.0,1532,2.0,...,No,Potential Lead,Mumbai,02.Medium,01.High,14.0,20.0,No,Yes,Email Opened
3,0cc2df48-7cf4-4e39-9de9-19797f9b38cc,660719,Landing Page Submission,Direct Traffic,No,No,0,1.0,305,1.0,...,No,Select,Mumbai,02.Medium,01.High,13.0,17.0,No,No,Modified
4,3256f628-e534-4826-9d63-4a8b88782852,660681,Landing Page Submission,Google,No,No,1,2.0,1428,1.0,...,No,Select,Mumbai,02.Medium,01.High,15.0,18.0,No,No,Modified


In [2]:
leads.shape

(9240, 37)

In [3]:
leads.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9240 entries, 0 to 9239
Data columns (total 37 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   Prospect ID                                    9240 non-null   object 
 1   Lead Number                                    9240 non-null   int64  
 2   Lead Origin                                    9240 non-null   object 
 3   Lead Source                                    9204 non-null   object 
 4   Do Not Email                                   9240 non-null   object 
 5   Do Not Call                                    9240 non-null   object 
 6   Converted                                      9240 non-null   int64  
 7   TotalVisits                                    9103 non-null   float64
 8   Total Time Spent on Website                    9240 non-null   int64  
 9   Page Views Per Visit                           9103 

In [4]:
leads.describe()

,Lead Number,Converted,TotalVisits,Total Time Spent on Website,Page Views Per Visit,Asymmetrique Activity Score,Asymmetrique Profile Score
count,9240.000000,9240.000000,9103.000000,9240.000000,9103.000000,5022.000000,5022.000000
mean,617188.435606,0.385390,3.445238,487.698268,2.362820,14.306252,16.344883
std,23405.995698,0.486714,4.854853,548.021466,2.161418,1.386694,1.811395
min,579533.000000,0.000000,0.000000,0.000000,0.000000,7.000000,11.000000
25%,596484.500000,0.000000,1.000000,12.000000,1.000000,14.000000,15.000000
50%,615479.000000,0.000000,3.000000,248.000000,2.000000,14.000000,16.000000
75%,637387.250000,1.000000,5.000000,936.000000,3.000000,15.000000,18.000000
max,660737.000000,1.000000,251.000000,2272.000000,55.000000,18.000000,20.000000


In [5]:
leads.describe(include='object')

,Prospect ID,Lead Origin,Lead Source,Do Not Email,Do Not Call,Last Activity,Country,Specialization,How did you hear about X Education,What is your current occupation,...,Lead Quality,Update me on Supply Chain Content,Get updates on DM Content,Lead Profile,City,Asymmetrique Activity Index,Asymmetrique Profile Index,I agree to pay the amount through cheque,A free copy of Mastering The Interview,Last Notable Activity
count,9240,9240,9204,9240,9240,9137,6779,7802,7033,6550,...,4473,9240,9240,6531,7820,5022,5022,9240,9240,9240
unique,9240,5,21,2,2,17,38,19,10,6,...,5,1,1,6,7,3,3,1,2,16
top,571b5c8e-a5b2-4d57-8574-f2ffb06fdeff,Landing Page Submission,Google,No,No,Email Opened,India,Select,Select,Unemployed,...,Might be,No,No,Select,Mumbai,02.Medium,02.Medium,No,No,Modified
freq,1,4886,2868,8506,9238,3437,6492,1942,5043,5600,...,1560,9240,9240,4146,3222,3839,2788,9240,6352,3407


In [6]:
round((leads.isnull().sum()/len(leads))*100,2)

Prospect ID                                       0.00
Lead Number                                       0.00
Lead Origin                                       0.00
Lead Source                                       0.39
Do Not Email                                      0.00
Do Not Call                                       0.00
Converted                                         0.00
TotalVisits                                       1.48
Total Time Spent on Website                       0.00
Page Views Per Visit                              1.48
Last Activity                                     1.11
Country                                          26.63
Specialization                                   15.56
How did you hear about X Education               23.89
What is your current occupation                  29.11
What matters most to you in choosing a course    29.32
Search                                            0.00
Magazine                                          0.00
Newspaper 

In [7]:
leads.isnull().sum()

Prospect ID                                         0
Lead Number                                         0
Lead Origin                                         0
Lead Source                                        36
Do Not Email                                        0
Do Not Call                                         0
Converted                                           0
TotalVisits                                       137
Total Time Spent on Website                         0
Page Views Per Visit                              137
Last Activity                                     103
Country                                          2461
Specialization                                   1438
How did you hear about X Education               2207
What is your current occupation                  2690
What matters most to you in choosing a course    2709
Search                                              0
Magazine                                            0
Newspaper Article           

In [8]:
(round((leads.isnull().sum()/
len(leads))*100,2)).sort_values(ascending=False)

Lead Quality                                     51.59
Asymmetrique Profile Index                       45.65
Asymmetrique Activity Score                      45.65
Asymmetrique Activity Index                      45.65
Asymmetrique Profile Score                       45.65
Tags                                             36.29
What matters most to you in choosing a course    29.32
Lead Profile                                     29.32
What is your current occupation                  29.11
Country                                          26.63
How did you hear about X Education               23.89
Specialization                                   15.56
City                                             15.37
Page Views Per Visit                              1.48
TotalVisits                                       1.48
Last Activity                                     1.11
Lead Source                                       0.39
Do Not Email                                      0.00
Lead Origi

In [9]:
cols = ['Prospect ID','Lead Number']
leads = leads.drop(cols,axis=1)

In [10]:
leads.shape

(9240, 35)

In [11]:
leads.shape

(9240, 35)

In [15]:
cols_to_drop = ['Lead Quality','Asymmetrique Profile Index','Asymmetrique Activity Score','Asymmetrique Activity Index','Asymmetrique Profile Score']
leads = leads.drop(cols_to_drop, axis=1)

KeyError: "['Lead Quality', 'Asymmetrique Profile Index', 'Asymmetrique Activity Score', 'Asymmetrique Activity Index', 'Asymmetrique Profile Score'] not found in axis"

In [16]:
leads.columns

Index(['Lead Origin', 'Lead Source', 'Do Not Email', 'Do Not Call',
       'Converted', 'TotalVisits', 'Total Time Spent on Website',
       'Page Views Per Visit', 'Last Activity', 'Country', 'Specialization',
       'How did you hear about X Education', 'What is your current occupation',
       'What matters most to you in choosing a course', 'Search', 'Magazine',
       'Newspaper Article', 'X Education Forums', 'Newspaper',
       'Digital Advertisement', 'Through Recommendations',
       'Receive More Updates About Our Courses', 'Tags',
       'Update me on Supply Chain Content', 'Get updates on DM Content',
       'Lead Profile', 'City', 'I agree to pay the amount through cheque',
       'A free copy of Mastering The Interview', 'Last Notable Activity'],
      dtype='object')

In [17]:
cols_to_drop = ['Lead Quality',
                'Asymmetrique Profile Index',
                'Asymmetrique Activity Score',
                'Asymmetrique Activity Index',
                'Asymmetrique Profile Score']

leads = leads.drop(cols_to_drop, axis=1, errors='ignore')

In [18]:
leads.shape

(9240, 30)

In [19]:
round((leads.isnull().sum()/
len(leads))*100,2).sort_values(ascending=False)

Tags                                             36.29
What matters most to you in choosing a course    29.32
Lead Profile                                     29.32
What is your current occupation                  29.11
Country                                          26.63
How did you hear about X Education               23.89
Specialization                                   15.56
City                                             15.37
Page Views Per Visit                              1.48
TotalVisits                                       1.48
Last Activity                                     1.11
Lead Source                                       0.39
Converted                                         0.00
Lead Origin                                       0.00
Total Time Spent on Website                       0.00
Do Not Email                                      0.00
Do Not Call                                       0.00
Search                                            0.00
Newspaper 

In [21]:
leads = leads[~pd.isnull(leads['Lead Profile'])]

In [22]:
leads.shape

(6531, 30)

In [23]:
round((leads.isnull().sum()/
len(leads))*100,2).sort_values(ascending=False)

Country                                          24.58
Tags                                             11.10
Page Views Per Visit                              1.99
TotalVisits                                       1.99
Last Activity                                     1.58
Lead Source                                       0.55
Converted                                         0.00
Do Not Call                                       0.00
Do Not Email                                      0.00
Total Time Spent on Website                       0.00
Specialization                                    0.00
How did you hear about X Education                0.00
What is your current occupation                   0.00
Lead Origin                                       0.00
What matters most to you in choosing a course     0.00
Search                                            0.00
Newspaper Article                                 0.00
Magazine                                          0.00
Newspaper 

In [24]:
leads = leads[~pd.isnull(leads['Country'])]

In [25]:
leads.shape

(4926, 30)

In [27]:
round((leads.isnull().sum()/len(leads))*100,2).sort_values(ascending=False)

Tags                                             7.94
Lead Source                                      0.02
Do Not Email                                     0.00
Do Not Call                                      0.00
Converted                                        0.00
Lead Origin                                      0.00
Total Time Spent on Website                      0.00
Page Views Per Visit                             0.00
Last Activity                                    0.00
Country                                          0.00
Specialization                                   0.00
How did you hear about X Education               0.00
What is your current occupation                  0.00
TotalVisits                                      0.00
What matters most to you in choosing a course    0.00
Search                                           0.00
Newspaper Article                                0.00
Magazine                                         0.00
Newspaper                   

In [28]:
leads = leads[~pd.isnull(leads['Tags'])]

In [29]:
leads.shape

(4535, 30)

In [30]:
round((leads.isnull().sum()/len(leads))*100,2).sort_values(ascending=False)

Lead Source                                      0.02
Lead Origin                                      0.00
Do Not Email                                     0.00
Do Not Call                                      0.00
Converted                                        0.00
TotalVisits                                      0.00
Total Time Spent on Website                      0.00
Page Views Per Visit                             0.00
Last Activity                                    0.00
Country                                          0.00
Specialization                                   0.00
How did you hear about X Education               0.00
What is your current occupation                  0.00
What matters most to you in choosing a course    0.00
Search                                           0.00
Magazine                                         0.00
Newspaper Article                                0.00
X Education Forums                               0.00
Newspaper                   

In [31]:
leads.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4535 entries, 1 to 9239
Data columns (total 30 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   Lead Origin                                    4535 non-null   object 
 1   Lead Source                                    4534 non-null   object 
 2   Do Not Email                                   4535 non-null   object 
 3   Do Not Call                                    4535 non-null   object 
 4   Converted                                      4535 non-null   int64  
 5   TotalVisits                                    4535 non-null   float64
 6   Total Time Spent on Website                    4535 non-null   int64  
 7   Page Views Per Visit                           4535 non-null   float64
 8   Last Activity                                  4535 non-null   object 
 9   Country                                        4535 non-n

In [32]:
varlist = ['Do Not Email','Do Not Call','Search','Magazine','Newspaper Article',
           'X Education Forums','Newspaper','Digital Advertisement',
           'Through Recommendations','Receive More Updates About Our Courses',
           'Update me on Supply Chain Content','Get updates on DM Content',
           'I agree to pay the amount through cheque',
           'A free copy of Mastering The Interview']

def binary_map(x):
    return x.map({'Yes':1,'No':0})

leads[varlist] = leads[varlist].apply(binary_map)

In [33]:
leads.head()

,Lead Origin,Lead Source,Do Not Email,Do Not Call,Converted,TotalVisits,Total Time Spent on Website,Page Views Per Visit,Last Activity,Country,...,Through Recommendations,Receive More Updates About Our Courses,Tags,Update me on Supply Chain Content,Get updates on DM Content,Lead Profile,City,I agree to pay the amount through cheque,A free copy of Mastering The Interview,Last Notable Activity
1,API,Organic Search,0,0,0,5.0,674,2.5,Email Opened,India,...,0,0,Ringing,0,0,Select,Select,0,0,Email Opened
2,Landing Page Submission,Direct Traffic,0,0,1,2.0,1532,2.0,Email Opened,India,...,0,0,Will revert after reading the email,0,0,Potential Lead,Mumbai,0,1,Email Opened
3,Landing Page Submission,Direct Traffic,0,0,0,1.0,305,1.0,Unreachable,India,...,0,0,Ringing,0,0,Select,Mumbai,0,0,Modified
4,Landing Page Submission,Google,0,0,1,2.0,1428,1.0,Converted to Lead,India,...,0,0,Will revert after reading the email,0,0,Select,Mumbai,0,0,Modified
6,Landing Page Submission,Google,0,0,1,2.0,1640,2.0,Email Opened,India,...,0,0,Will revert after reading the email,0,0,Potential Lead,Mumbai,0,0,Modified


In [34]:
dummy = pd.get_dummies(leads[['Lead Origin','Lead Source']], drop_first=True)

In [35]:
dummy.head()

,Lead Origin_Landing Page Submission,Lead Origin_Lead Add Form,Lead Source_Facebook,Lead Source_Google,Lead Source_Olark Chat,Lead Source_Organic Search,Lead Source_Pay per Click Ads,Lead Source_Press_Release,Lead Source_Reference,Lead Source_Referral Sites,Lead Source_Social Media,Lead Source_WeLearn,Lead Source_Welingak Website,Lead Source_bing,Lead Source_testone
1,False,False,False,False,False,True,False,False,False,False,False,False,False,False,False
2,True,False,False,False,False,False,False,False,False,False,False,False,False,False,False
3,True,False,False,False,False,False,False,False,False,False,False,False,False,False,False
4,True,False,False,True,False,False,False,False,False,False,False,False,False,False,False
6,True,False,False,True,False,False,False,False,False,False,False,False,False,False,False


In [37]:
leads = pd.concat([leads, dummy], axis=1)

In [38]:
leads.drop(['Lead Origin','Lead Source'], axis=1, inplace=True)

In [39]:
leads.head()

,Do Not Email,Do Not Call,Converted,TotalVisits,Total Time Spent on Website,Page Views Per Visit,Last Activity,Country,Specialization,How did you hear about X Education,...,Lead Source_Organic Search,Lead Source_Pay per Click Ads,Lead Source_Press_Release,Lead Source_Reference,Lead Source_Referral Sites,Lead Source_Social Media,Lead Source_WeLearn,Lead Source_Welingak Website,Lead Source_bing,Lead Source_testone
1,0,0,0,5.0,674,2.5,Email Opened,India,Select,Select,...,True,False,False,False,False,False,False,False,False,False
2,0,0,1,2.0,1532,2.0,Email Opened,India,Business Administration,Select,...,False,False,False,False,False,False,False,False,False,False
3,0,0,0,1.0,305,1.0,Unreachable,India,Media and Advertising,Word Of Mouth,...,False,False,False,False,False,False,False,False,False,False
4,0,0,1,2.0,1428,1.0,Converted to Lead,India,Select,Other,...,False,False,False,False,False,False,False,False,False,False
6,0,0,1,2.0,1640,2.0,Email Opened,India,Supply Chain Management,Online Search,...,False,False,False,False,False,False,False,False,False,False


In [40]:
varlist = ['Specialization','What is your current occupation','What matters most to you in choosing a course']

In [41]:
dummy = pd.get_dummies(leads[varlist], drop_first=True)

In [42]:
leads = pd.concat([leads, dummy], axis=1)

In [43]:
leads.drop(varlist, axis=1, inplace=True)

In [44]:
leads.head()

,Do Not Email,Do Not Call,Converted,TotalVisits,Total Time Spent on Website,Page Views Per Visit,Last Activity,Country,How did you hear about X Education,Search,...,Specialization_Select,Specialization_Services Excellence,Specialization_Supply Chain Management,Specialization_Travel and Tourism,What is your current occupation_Housewife,What is your current occupation_Other,What is your current occupation_Student,What is your current occupation_Unemployed,What is your current occupation_Working Professional,What matters most to you in choosing a course_Other
1,0,0,0,5.0,674,2.5,Email Opened,India,Select,0,...,True,False,False,False,False,False,False,True,False,False
2,0,0,1,2.0,1532,2.0,Email Opened,India,Select,0,...,False,False,False,False,False,False,True,False,False,False
3,0,0,0,1.0,305,1.0,Unreachable,India,Word Of Mouth,0,...,False,False,False,False,False,False,False,True,False,False
4,0,0,1,2.0,1428,1.0,Converted to Lead,India,Other,0,...,True,False,False,False,False,False,False,True,False,False
6,0,0,1,2.0,1640,2.0,Email Opened,India,Online Search,0,...,False,False,True,False,False,False,False,True,False,False


In [45]:
leads.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4535 entries, 1 to 9239
Data columns (total 64 columns):
 #   Column                                                Non-Null Count  Dtype  
---  ------                                                --------------  -----  
 0   Do Not Email                                          4535 non-null   int64  
 1   Do Not Call                                           4535 non-null   int64  
 2   Converted                                             4535 non-null   int64  
 3   TotalVisits                                           4535 non-null   float64
 4   Total Time Spent on Website                           4535 non-null   int64  
 5   Page Views Per Visit                                  4535 non-null   float64
 6   Last Activity                                         4535 non-null   object 
 7   Country                                               4535 non-null   object 
 8   How did you hear about X Education                    4535 non-

In [46]:
varlist = ['Last Activity','Country','Lead Profile']

In [47]:
dummy = pd.get_dummies(leads[varlist], drop_first=True)

In [48]:
leads = pd.concat([leads, dummy], axis=1)

In [49]:
leads.drop(varlist, axis=1, inplace=True)

In [50]:
leads.head()

,Do Not Email,Do Not Call,Converted,TotalVisits,Total Time Spent on Website,Page Views Per Visit,How did you hear about X Education,Search,Magazine,Newspaper Article,...,Country_Uganda,Country_United Arab Emirates,Country_United Kingdom,Country_United States,Country_Vietnam,Lead Profile_Lateral Student,Lead Profile_Other Leads,Lead Profile_Potential Lead,Lead Profile_Select,Lead Profile_Student of SomeSchool
1,0,0,0,5.0,674,2.5,Select,0,0,0,...,False,False,False,False,False,False,False,False,True,False
2,0,0,1,2.0,1532,2.0,Select,0,0,0,...,False,False,False,False,False,False,False,True,False,False
3,0,0,0,1.0,305,1.0,Word Of Mouth,0,0,0,...,False,False,False,False,False,False,False,False,True,False
4,0,0,1,2.0,1428,1.0,Other,0,0,0,...,False,False,False,False,False,False,False,False,True,False
6,0,0,1,2.0,1640,2.0,Online Search,0,0,0,...,False,False,False,False,False,False,False,True,False,False


In [51]:
leads.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4535 entries, 1 to 9239
Columns: 112 entries, Do Not Email to Lead Profile_Student of SomeSchool
dtypes: bool(90), float64(2), int64(16), object(4)
memory usage: 1.2+ MB


In [52]:
leads.select_dtypes(include=['object']).columns

Index(['How did you hear about X Education', 'Tags', 'City',
       'Last Notable Activity'],
      dtype='object')

In [53]:
varlist = ['How did you hear about X Education',
           'Tags',
           'City',
           'Last Notable Activity']

In [54]:
dummy = pd.get_dummies(leads[varlist], drop_first=True)

In [55]:
leads = pd.concat([leads, dummy], axis=1)

In [56]:
leads.drop(varlist, axis=1, inplace=True)

In [57]:
leads.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4535 entries, 1 to 9239
Columns: 161 entries, Do Not Email to Last Notable Activity_View in browser link Clicked
dtypes: bool(143), float64(2), int64(16)
memory usage: 1.3 MB


In [58]:
leads.select_dtypes(include=['object']).columns

Index([], dtype='object')

In [59]:
from sklearn.model_selection import train_test_split

In [60]:
X = leads.drop('Converted', axis=1)
y = leads['Converted']

In [61]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=100
)

In [62]:
from sklearn.preprocessing import StandardScaler

In [63]:
scaler = StandardScaler()

In [65]:
X_train[['TotalVisits','Total Time Spent on Website','Page Views Per Visit']] = scaler.fit_transform(
    X_train[['TotalVisits','Total Time Spent on Website','Page Views Per Visit']]
)

In [66]:
import statsmodels.api as sm

In [67]:
logm1 = sm.GLM(y_train, sm.add_constant(X_train), family=sm.families.Binomial())

ValueError: Pandas data cast to numpy dtype of object. Check input data with np.asarray(data).

In [68]:
X_train = X_train.astype(int)

In [69]:
logm1 = sm.GLM(y_train, sm.add_constant(X_train), family=sm.families.Binomial())

In [70]:
res = logm1.fit()

In [71]:
res.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:              Converted   No. Observations:                 3174
Model:                            GLM   Df Residuals:                     3036
Model Family:                Binomial   Df Model:                          137
Link Function:                  Logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -357.63
Date:                Sat, 23 May 2026   Deviance:                       715.27
Time:                        11:46:37   Pearson chi2:                 2.57e+03
No. Iterations:                    28   Pseudo R-squ. (CS):             0.6799
Covariance Type:            nonrobust                                         
============================================================================================================================
                                                               coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------------------------------------------------
const                                                       46.3902   1.89e+06   2.46e-05      1.000    -3.7e+06     3.7e+06
Do Not Email                                                -0.5955      0.614     -0.969      0.332      -1.800       0.609
Do Not Call                                                 24.9962   3.54e+05   7.06e-05      1.000   -6.94e+05    6.94e+05
TotalVisits                                                  0.2398      0.115      2.078      0.038       0.014       0.466
Total Time Spent on Website                                  1.3167      0.181      7.267      0.000       0.962       1.672
Page Views Per Visit                                        -0.1959      0.169     -1.162      0.245      -0.526       0.135
Search                                                      -0.9119      3.394     -0.269      0.788      -7.565       5.741
Magazine                                                  6.621e-08      0.072   9.19e-07      1.000      -0.141       0.141
Newspaper Article                                           18.7892   1.17e+04      0.002      0.999    -2.3e+04     2.3e+04
X Education Forums                                       -1.043e-07      0.097  -1.07e-06      1.000      -0.191       0.191
Newspaper                                                -1.609e-07      0.185   -8.7e-07      1.000      -0.363       0.363
Digital Advertisement                                      -31.0765   3.28e+07  -9.47e-07      1.000   -6.43e+07    6.43e+07
Through Recommendations                                     25.7480   9.71e+05   2.65e-05      1.000    -1.9e+06     1.9e+06
Receive More Updates About Our Courses                     8.49e-08      0.088   9.62e-07      1.000      -0.173       0.173
Update me on Supply Chain Content                         3.624e-07      0.378   9.59e-07      1.000      -0.740       0.740
Get updates on DM Content                                -2.104e-07      0.195  -1.08e-06      1.000      -0.381       0.381
I agree to pay the amount through cheque                 -1.731e-07      0.181  -9.58e-07      1.000      -0.354       0.354
A free copy of Mastering The Interview                       0.0113      0.373      0.030      0.976      -0.720       0.743
Lead Origin_Landing Page Submission                          0.2818      0.471      0.599      0.549      -0.641       1.205
Lead Origin_Lead Add Form                                   18.7974   1.27e+06   1.48e-05      1.000   -2.49e+06    2.49e+06
Lead Source_Facebook                                     -1.388e-07      0.152  -9.15e-07      1.000      -0.297       0.297
Lead Source_Google                                           0.1836      0.348     

In [72]:
res.summary2()

<class 'statsmodels.iolib.summary2.Summary'>
"""
                                             Results: Generalized linear model
===========================================================================================================================
Model:                                   GLM                                 AIC:                               991.2682   
Link Function:                           Logit                               BIC:                               -23763.2344
Dependent Variable:                      Converted                           Log-Likelihood:                    -357.63    
Date:                                    2026-05-23 11:47                    LL-Null:                           -2165.4    
No. Observations:                        3174                                Deviance:                          715.27     
Df Model:                                137                                 Pearson chi2:                      2.57e+03   
Df Residuals:                            3036                                Scale:                             1.0000     
Method:                                  IRLS                                                                              
---------------------------------------------------------------------------------------------------------------------------
                                                          Coef.      Std.Err.      z    P>|z|      [0.025         0.975]   
---------------------------------------------------------------------------------------------------------------------------
const                                                     46.3902  1887960.1654  0.0000 1.0000  -3700287.5381  3700380.3186
Do Not Email                                              -0.5955        0.6144 -0.9692 0.3325        -1.7998        0.6088
Do Not Call                                               24.9962   354216.6391  0.0001 0.9999   -694226.8592   694276.8516
TotalVisits                                                0.2398        0.1154  2.0778 0.0377         0.0136        0.4659
Total Time Spent on Website                                1.3167        0.1812  7.2666 0.0000         0.9615        1.6718
Page Views Per Visit                                      -0.1959        0.1687 -1.1615 0.2454        -0.5265        0.1347
Search                                                    -0.9119        3.3943 -0.2687 0.7882        -7.5647        5.7409
Magazine                                                   0.0000        0.0721  0.0000 1.0000        -0.1413        0.1413
Newspaper Article                                         18.7892    11722.1337  0.0016 0.9987    -22956.1707    22993.7492
X Education Forums                                        -0.0000        0.0974 -0.0000 1.0000        -0.1909        0.1909
Newspaper                                                 -0.0000        0.1850 -0.0000 1.0000        -0.3626        0.3626
Digital Advertisement                                    -31.0765 32799329.0398 -0.0000 1.0000 -64285534.7116 64285472.5586
Through Recommendations                                   25.7480   970962.0739  0.0000 1.0000  -1903024.9471  1903076.4432
Receive More Updates About Our Courses                     0.0000        0.0883  0.0000 1.0000        -0.1730        0.1730
Update me on Supply Chain Content                          0.0000        0.3777  0.0000 1.0000        -0.7403        0.7403
Get updates on DM Content                                 -0.0000        0.1946 -0.0000 1.0000        -0.3813        0.3813
I agree to pay the amount through cheque                  -0.0000        0.1807 -0.0000 1.0000        -0.3541        0.3541
A free copy of Mastering The Interview                     0.0113        0.3732  0.0303 0.9758        -0.7201        0.7427
Lead Origin_Landing Page Submission                        0.2818        0.4709  0.5985 0.5495        -0.6411        1.2048
Lead Origin_Lead Add Form   

In [73]:
res.pvalues.sort_values(ascending=False).head(10)

Country_China                                    1.000000
Country_Italy                                    1.000000
Last Activity_Visited Booth in Tradeshow         0.999999
Last Notable Activity_Email Link Clicked         0.999999
Country_Kenya                                    0.999999
Country_Tanzania                                 0.999999
Lead Source_testone                              0.999999
Newspaper                                        0.999999
Last Notable Activity_Page Visited on Website    0.999999
Lead Source_Facebook                             0.999999
dtype: float64

In [74]:
col = ['Country_China',
       'Country_Italy',
       'Last Activity_Visited Booth in Tradeshow',
       'Last Notable Activity_Email Link Clicked',
       'Country_Kenya',
       'Country_Tanzania',
       'Lead Source_testone',
       'Newspaper',
       'Last Notable Activity_Page Visited on Website',
       'Lead Source_Facebook']

X_train = X_train.drop(col, axis=1)

In [75]:
logm1 = sm.GLM(y_train, sm.add_constant(X_train), family=sm.families.Binomial())
res = logm1.fit()
res.summary2()

C:\Users\Dell\anaconda3\Lib\site-packages\statsmodels\genmod\families\links.py:198: RuntimeWarning: overflow encountered in exp
  t = np.exp(-z)


MemoryError: 

In [1]:
from sklearn.linear_model import LogisticRegression

logreg = LogisticRegression(max_iter=1000)
logreg.fit(X_train, y_train)

NameError: name 'X_train' is not defined

In [5]:
leads.shape

(9240, 37)

In [6]:
leads = leads.drop(['Prospect ID','Lead Number'], axis=1)

In [7]:
leads.shape

(9240, 35)

In [8]:
leads.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9240 entries, 0 to 9239
Data columns (total 35 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   Lead Origin                                    9240 non-null   object 
 1   Lead Source                                    9204 non-null   object 
 2   Do Not Email                                   9240 non-null   object 
 3   Do Not Call                                    9240 non-null   object 
 4   Converted                                      9240 non-null   int64  
 5   TotalVisits                                    9103 non-null   float64
 6   Total Time Spent on Website                    9240 non-null   int64  
 7   Page Views Per Visit                           9103 non-null   float64
 8   Last Activity                                  9137 non-null   object 
 9   Country                                        6779 

In [9]:
round(100*(leads.isnull().sum()/len(leads.index)),2).sort_values(ascending=False)

Lead Quality                                     51.59
Asymmetrique Activity Index                      45.65
Asymmetrique Activity Score                      45.65
Asymmetrique Profile Score                       45.65
Asymmetrique Profile Index                       45.65
Tags                                             36.29
What matters most to you in choosing a course    29.32
Lead Profile                                     29.32
What is your current occupation                  29.11
Country                                          26.63
How did you hear about X Education               23.89
Specialization                                   15.56
City                                             15.37
Page Views Per Visit                              1.48
TotalVisits                                       1.48
Last Activity                                     1.11
Lead Source                                       0.39
Do Not Email                                      0.00
Lead Origi

In [10]:
cols = ['Lead Quality',
        'Asymmetrique Activity Index',
        'Asymmetrique Activity Score',
        'Asymmetrique Profile Index',
        'Asymmetrique Profile Score',
        'Tags']

leads = leads.drop(cols, axis=1)

In [11]:
leads.shape

(9240, 29)

In [13]:
leads = leads.dropna()

In [14]:
leads.shape

(4925, 29)

In [15]:
round(100*(leads.isnull().sum()/len(leads.index)),2).sort_values(ascending=False)

Lead Origin                                      0.0
Lead Source                                      0.0
Do Not Email                                     0.0
Do Not Call                                      0.0
Converted                                        0.0
TotalVisits                                      0.0
Total Time Spent on Website                      0.0
Page Views Per Visit                             0.0
Last Activity                                    0.0
Country                                          0.0
Specialization                                   0.0
How did you hear about X Education               0.0
What is your current occupation                  0.0
What matters most to you in choosing a course    0.0
Search                                           0.0
Magazine                                         0.0
Newspaper Article                                0.0
X Education Forums                               0.0
Newspaper                                     

In [16]:
leads.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4925 entries, 1 to 9239
Data columns (total 29 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   Lead Origin                                    4925 non-null   object 
 1   Lead Source                                    4925 non-null   object 
 2   Do Not Email                                   4925 non-null   object 
 3   Do Not Call                                    4925 non-null   object 
 4   Converted                                      4925 non-null   int64  
 5   TotalVisits                                    4925 non-null   float64
 6   Total Time Spent on Website                    4925 non-null   int64  
 7   Page Views Per Visit                           4925 non-null   float64
 8   Last Activity                                  4925 non-null   object 
 9   Country                                        4925 non-n

In [17]:
varlist = ['Lead Origin',
           'Lead Source',
           'Do Not Email',
           'Do Not Call',
           'Last Activity',
           'Country',
           'Specialization',
           'How did you hear about X Education',
           'What is your current occupation',
           'What matters most to you in choosing a course',
           'Search',
           'Magazine',
           'Newspaper Article',
           'X Education Forums',
           'Newspaper',
           'Digital Advertisement',
           'Through Recommendations',
           'Receive More Updates About Our Courses',
           'Update me on Supply Chain Content',
           'Get updates on DM Content',
           'Lead Profile',
           'City',
           'I agree to pay the amount through cheque',
           'A free copy of Mastering The Interview',
           'Last Notable Activity']

In [18]:
dummy = pd.get_dummies(leads[varlist], drop_first=True)

In [19]:
leads = pd.concat([leads, dummy], axis=1)

In [20]:
leads = leads.drop(varlist, axis=1)

In [21]:
leads.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4925 entries, 1 to 9239
Columns: 132 entries, Converted to Last Notable Activity_View in browser link Clicked
dtypes: bool(128), float64(2), int64(2)
memory usage: 808.0 KB


In [22]:
X = leads.drop('Converted', axis=1)
y = leads['Converted']

In [23]:
from sklearn.model_selection import train_test_split

In [24]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=100
)

In [25]:
from sklearn.preprocessing import StandardScaler

In [26]:
scaler = StandardScaler()

In [27]:
X_train[['TotalVisits','Total Time Spent on Website','Page Views Per Visit']] = scaler.fit_transform(
    X_train[['TotalVisits','Total Time Spent on Website','Page Views Per Visit']]
)

In [28]:
X_train[['TotalVisits','Total Time Spent on Website','Page Views Per Visit']] = scaler.fit_transform(
    X_train[['TotalVisits','Total Time Spent on Website','Page Views Per Visit']]
)

In [32]:
logreg.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

In [33]:
y_pred = logreg.predict(X_test)

In [34]:
from sklearn.metrics import accuracy_score

In [35]:
accuracy_score(y_test, y_pred)

0.4506089309878214

In [36]:
from sklearn.metrics import classification_report

In [37]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.00      0.00       813
           1       0.45      1.00      0.62       665

    accuracy                           0.45      1478
   macro avg       0.73      0.50      0.31      1478
weighted avg       0.75      0.45      0.28      1478



In [38]:
logreg = LogisticRegression(max_iter=1000, class_weight='balanced')

In [39]:
logreg.fit(X_train, y_train)

LogisticRegression(class_weight='balanced', max_iter=1000)

In [40]:
y_pred = logreg.predict(X_test)

In [41]:
accuracy_score(y_test, y_pred)

0.44993234100135315

In [42]:
from sklearn.metrics import confusion_matrix

In [43]:
confusion_matrix(y_test, y_pred)

array([[  0, 813],
       [  0, 665]])

Conclusion:
The Logistic Regression model was successfully implemented on the Leads dataset.
Data preprocessing techniques such as handling missing values, encoding categorical variables, scaling and train-test splitting were performed sucessfully.
The model achieved an accuracy of approximately 45% on the test data.
Model evaluation was carried out using accuracy score, classification report, and confusion matrix.
